client/src/components/game/gui/panels/chat/ChatPanel.jsx

```diff
@@ -0,0 +1,69 @@
+import React, { useState, useRef, useEffect } from "react";
+import PropTypes from "prop-types";
+import "./ChatPanel.scss";
+import PubSub from "pubsub-js";
+import { ApplicationState, GUIState, PlayerState } from "../../../../../shared/state/States";
+import { PANEL_CHANGE } from "../../../../../shared/EventTypes";
+import Panels from "../PanelsEnum";
+
+function ChatPanel({ onCloseCallback }) {
+    const [chats, sendChat] = useState([]);
+    const chatContentsRef = useRef(null);
+    const chatInputRef = useRef(null);
+
+    useEffect(() => {
+        const subs = [
+            PubSub.subscribe(PANEL_CHANGE, () => {
+                console.log("panel change from chats");
+                if (GUIState.activePanel === Panels.Chat) {
+                    chatInputRef.current.focus();
+                }
+            }),
+        ];
+
+        return () => {
+            subs.forEach((sub) => {
+                PubSub.unsubscribe(sub);
+            });
+        };
+    });
+
+    const scrollChatToBottom = () => {
+        chatContentsRef.current.scrollTop = chatContentsRef.current.scrollHeight;
+    };
+
+    const sendC = (e) => {
+        if (e.key === "Enter") {
+            const text = e.target.value;
+            if (!text) return;
+
+            const newChat = `${PlayerState.displayName}: ${text}`;
+            sendChat((prevChats) => [...prevChats, newChat]);
+
+            ApplicationState.connection.sendEvent("chat", text);
+
+            e.target.value = "";
+
+            // add some delay to properly scroll down to edge of chats
+            setTimeout(scrollChatToBottom, 10);
+        }
+    };
+
+    return (
+        <div className="chat-container gui-zoomable">
+            <div className="chat-contents" ref={chatContentsRef}>
+                { chats.map((chat) => (<p className="chat-text" key={chat + Math.random() * 100000}>{chat}</p>)) }
```

Comment:
And I would recommend moving these chat messages lines out into some `ChatLine` component, with separate states for the display name and message content, so they can be styled a bit differently, like they are for the input below, where the display name is bold. Something like:
```js
{ chats.map((chat) => (
    <ChatLine
        key={chat.key}
        displayName={chat.displayName}
        message={chat.message}
    />
)) }
```